### ISSUE TITLE
Segfault: Exception thrown during bootstrappingBasics)

### ISSUE THREAD

**Issue by tahpot** [preillyme/v8js#136]
I have a repeatable segfault when attempting to load the Handlebars.js library.

The segfault can be repeated with the following code:

```
$handlebarsJs = "var root = typeof global !== 'undefined' ? global : window, \$Handlebars = 'test';";
V8Js::registerExtension('handlebars', $handlebarsJs, [], true);
```

The full error message is:
```
Exception thrown during bootstrappingBasics) ... 
Extension or internal compilation error in handlebars at line 2.
Error installing extension 'handlebars'.
Segmentation fault: 11
```

Running the same code in v8 works fine:
```
$ v8
V8 version 3.30.33.16 [sample shell]
> var root = typeof global !== 'undefined' ? global : window, $Handlebars = 'test';
(shell):1: ReferenceError: window is not defined
var root = typeof global !== 'undefined' ? global : window, $Handlebars = 'tes
                                                    ^
ReferenceError: window is not defined
    at (shell):1:53
> window = null;
null
> var root = typeof global !== 'undefined' ? global : window, $Handlebars = 'test';
> ^D 
```

**Comment by guilhermeblanco** [comment]
I get the same error... basically, any undefined variable that creates either an Error (TypeError, ReferenceError, etc) will cause a segfault. =(

**Comment by rosmo** [comment]
Can you get a gdb backtrace from the process?

**Comment by pinepain** [comment]
The whole problem is leaving `while (!V8JSG(timer_stop))` loop in [`void v8js_timer_thread(TSRMLS_D)`](https://github.com/preillyme/v8js/blob/2954de2db50b0a29a6e31fb3bceef6be3074498f/v8js_timer.cc#L67) function (this function runs in separate thread). When we leave that `while` loop, we literally exit the `v8js_timer_thread` function, so literally we finish threaded logic execution and the next good thing we can do is to `join` thread and clean all thread-related data structures. So at the end of that function it is nice to have something like

```
V8JSG(timer_thread)->join();
delete V8JSG(timer_thread);
V8JSG(timer_thread) = NULL;
```

which should solve the problem. In real life this code should be placed in [`v8js_v8_call`](https://github.com/preillyme/v8js/blob/90c3b9db5f7f3ca9adf0ddbdcc86fe2893a42e7d/v8js_v8.cc#L123) with `v8js_ctx.in_execution == 0` value check or something like this.

While this is a solution, it is a bit dirty and it looks like whole things with thread starting and destroying was designed to start thread only once and then just populate/cleanup `V8JSG(timer_stack)`. So I can assume that the intention is to move `!V8JSG(timer_stop)` check under `while` loop to have logic like this:

```
while(!V8JSG(request_shutdown)) {
    if (!V8JSG(timer_stop)) {
        // loop over V8JSG(timer_stack)
    }
    // sleep for some time
}
```

and then, say, in PHP request shutdown handler (that one that has `PHP_RSHUTDOWN_FUNCTION` signature) write something like:

```
V8JSG(request_shutdown) = true;
V8JSG(timer_thread)->join();
delete V8JSG(timer_thread);
V8JSG(timer_thread) = NULL;
```

**Comment by rosmo** [comment]
It's not too dirty in my opinion. Mind if you submit a PR?

**Comment by pinepain** [comment]
Which one, first with thread interrupting after js script execution done, or the second one - thread interrupting when request done?

**Comment by rosmo** [comment]
I think the rshutdown function is not the best place, since you can have long running PHP cli processes which use V8Js. So the first one is better in my opinion.

**Comment by pinepain** [comment]
I will play more with this in the evening, so if no one has better ideas I will PR later, say ETA up to 1 day.

**Comment by pinepain** [comment]
The discussion above (about threading and timer) is about [issue #138](https://github.com/preillyme/v8js/issues/138). I miss the tabs with issues. Sorry.

**Comment by pinepain** [comment]
The discussion above (about threading and timer) is about [issue #138](https://github.com/preillyme/v8js/issues/138). I miss the tabs with issues. Sorry.
